src/jit/function-builder.cc

```diff
@@ -246,116 +246,145 @@ bool FunctionBuilder::buildIL() {
   setVMState(new TR::VirtualMachineState());
 
   const uint8_t* istream = thread_->GetIstream();
+  const uint8_t* pc = &istream[fn_->offset];
+  auto* state = new WabtState();
 
-  workItems_.emplace_back(OrphanBytecodeBuilder(0, const_cast<char*>(ReadOpcodeAt(&istream[fn_->offset]).GetName())),
-                          &istream[fn_->offset]);
+  SetUpLocals(this, &pc, &state->stack);
+  setVMState(state);
+
+  workItems_.emplace_back(OrphanBytecodeBuilder(0, const_cast<char*>(ReadOpcodeAt(pc).GetName())),
+                          pc);
   AppendBuilder(workItems_[0].builder);
 
   int32_t next_index;
 
   while ((next_index = GetNextBytecodeFromWorklist()) != -1) {
     auto& work_item = workItems_[next_index];
 
-    if (!Emit(work_item.builder, istream, work_item.pc))
+    if (!Emit(work_item.builder, static_cast<WabtState*>(work_item.builder->vmState())->stack, istream, work_item.pc))
       return false;
   }
 
   return true;
 }
 
-/**
- * @brief Generate push to the interpreter stack
- *
- * The generated code should be equivalent to:
- *
- * auto stack_top = *stack_top_addr;
- * stack_base_addr[stack_top] = value;
- * *stack_top_addr = stack_top + 1;
- */
-void FunctionBuilder::Push(TR::IlBuilder* b, const char* type, TR::IlValue* value, const uint8_t* pc) {
+void FunctionBuilder::SetUpLocals(TR::IlBuilder* b, const uint8_t** pc, VirtualStack* stack) {
+  // Add placeholders onto the virtual stack. These values should never be actually read, but will
+  // eventually be dropped using drop or drop_keep in the function epilogue.
+  for (size_t i = 0; i < fn_->param_and_local_types.size(); i++) {
+    stack->Push(nullptr);
+  }
+
+  if (fn_->local_count == 0) return;
+
+  Opcode opcode = ReadOpcode(pc);
+  TR_ASSERT_FATAL(opcode == Opcode::InterpAlloca, "Function with locals is missing alloca");
+  TR_ASSERT_FATAL(ReadU32(pc) == fn_->local_count, "Function has wrong alloca size");
```

Comment:
Can the expected and actual alloca size be printed as part of the assert message?